src/client.cpp

```diff
@@ -280,6 +290,36 @@ void CClient::OnNewConnection()
     //### TODO: END ###//
 }
 
+void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
+{
+    // Upon receiving a new client list, we have to reset oldGain and newGain
+    // entries for unused channels. This ensures that a disconnected channel
+    // does not leave behind wrong cached gain values which would leak into
+    // any new channel which reused this channel id.
+    int  iNumConnectedClients = vecChanInfo.Size();
+    bool bChanIdInUse;
+    for ( int iId = 0; iId < MAX_NUM_CHANNELS; iId++ )
+    {
+        bChanIdInUse = false;
+        for ( int j = 0; j < iNumConnectedClients; j++ )
+        {
+            if ( vecChanInfo[j].iChanID == iId )
+            {
+                // channel was there and still is. therefore, keep the
+                // value in oldGain.
+                bChanIdInUse = true;
+                break;
+            }
+        }
+        if ( !bChanIdInUse )
+        {
+            // reset oldGain and newGain as this channel id is currently unused and will
+            // start with a server-side gain at 1 (100%) again.
+            oldGain[iId] = newGain[iId] = 1;
+        }
+    }
+}
```

Comment:
```suggestion
void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
{
    // Upon receiving a new client list, we have to reset oldGain and newGain
    // entries for unused channels. This ensures that a disconnected channel
    // does not leave behind wrong cached gain values which would leak into
    // any new channel which reused this channel id.

    int iNumConnectedClients = vecChanInfo.Size();

    bool inUse[MAX_NUM_CHANNELS];

    // get all channels which are in use from vacChanInfo[]
    for ( int i = 0; i < iNumConnectedClients; i++ )
    {
        inUse[vecChanInfo[i].iChanID] = true;
    }

    // set all channels which are not in use to gain 1
    for ( int i = 0; i < MAX_NUM_CHANNELS; i++ )
    {
        if ( !inUse[i] )
        {
            // reset oldGain and newGain as this channel id is currently unused and will
            // start with a server-side gain at 1 (100%) again.
            oldGain[i] = newGain[i] = 1;
        }
    }
}
```

Reply:
I think it solves the bug. However, I'd like an independent test of the code from above.

Reply:
Thanks. Based on your earlier suggestions I came up with almost the same. I've just pushed that with proper attribution.